monitoring/asset_proof_collector.go

```diff
@@ -0,0 +1,100 @@
+package monitoring
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	numAssetProofsMetric = "asset_proofs"
+)
+
+// assetProofCollector is a Prometheus collector that exports proof related data
+// of all taproot assets.
+type assetProofCollector struct {
+	collectMx sync.Mutex
+
+	cfg      *PrometheusConfig
+	registry *prometheus.Registry
+
+	proofsVec *prometheus.GaugeVec
+}
+
+func newAssetProofCollector(cfg *PrometheusConfig,
+	registry *prometheus.Registry) (*assetProofCollector, error) {
+
+	if cfg == nil {
+		return nil, errors.New("asset proof collector prometheus cfg " +
+			"is nil")
+	}
+
+	if cfg.AssetStore == nil {
+		return nil, errors.New("asset proof collector asset store is " +
+			"nil")
+	}
+
+	return &assetProofCollector{
+		cfg:      cfg,
+		registry: registry,
+		proofsVec: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: numAssetProofsMetric,
+				Help: "Proofs of all assets",
+			},
+			[]string{"asset_name"},
+		),
+	}, nil
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once the
+// last descriptor has been sent.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *assetProofCollector) Describe(ch chan<- *prometheus.Desc) {
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	a.proofsVec.Describe(ch)
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *assetProofCollector) Collect(ch chan<- prometheus.Metric) {
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	ctxdb, cancel := context.WithTimeout(context.Background(), dbTimeout)
+	defer cancel()
+
+	// Fetch all proofs.
+	proofs, err := a.cfg.AssetStore.FetchAssetProofs(ctxdb)
```

Comment:
With the default scraping interval, I think this'll run every 60 seconds or so. Extrapolating out to say 100x the current scale, will this be fast enough?

We may want to consider a more precise query than what we have here.

Reply:
Agree

If we want to stick with taking a snapshot of **all** asset proofs and keep track of them, we don't have anything better currently. I believe we can stick to this for now, and our loadtests will indicate the next step.

Also, if we want to keep the loadtest behavior "random" (i.e send a random asset X times) we can't really know which asset is going to be used, so it kind of makes sense to keep track of everything.

If we change the loadtest behavior so that specific assets are being sent around in specific patterns, then we can change the above query and only track what we're interested in